prow/flagutil/config.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flagutil
+
+import (
+	"flag"
+
+	"k8s.io/test-infra/prow/config"
+)
+
+// ConfigOptions holds options for interacting with Configs.
+type ConfigOptions struct {
+	ConfigPath    string
+	jobConfigPath string
+}
+
+// AddFlags injects config options into the given FlagSet.
+func (o *ConfigOptions) AddFlags(fs *flag.FlagSet) {
+	fs.StringVar(&o.ConfigPath, "config-path", "/etc/config/config.yaml", "Path to config.yaml.")
+	fs.StringVar(&o.jobConfigPath, "job-config-path", "", "Path to prow job configs.")
+}
+
+// Validate validates config options.
+func (o *ConfigOptions) Validate(dryRun bool) error {
+	return nil
+}
+
+// Agent returns a config agent.
+func (o *ConfigOptions) Agent(start bool) (agent *config.Agent, err error) {
+	agent = &config.Agent{}
+	config, err := config.Load(o.ConfigPath, o.jobConfigPath)
+	if err != nil {
+		return nil, err
+	}
+	agent.Set(config)
+
+	if start {
```

Comment:
This seems weird -- in what cases do you want an unstarted agent. Or, why not always return one either ready for use or always unready?

Reply:
> ... in what cases do you want an unstarted agent.

In [this one][1].  Or at least master isn't starting there.

> Or, why not always return one either ready for use or always unready?

Passing a boolean `start` is easier than two separate calls with error checks, and most consumers want a started agent.

[1]: https://github.com/kubernetes/test-infra/pull/9701/commits/eb7c2eda8f88e13092e5560fab76c741766912e5#diff-4b382ccce825880cc0a4c58c43b8d8e6R91

Reply:
Ah. Those are CLI tools that run once, not as daemons, so using the config agent wasn't useful in the past. I think it may be better to just always start the agent and in those cases, continue to read the config just once. It won't matter that the agent is running in the background while the tool does whatever else and it will clean up this interface for all callers.

Reply:
> I think it may be better to just always start the agent and in those cases, continue to read the config just once.

[Done][1].

[1]: https://github.com/kubernetes/test-infra/pull/9701#issuecomment-430418425